cmd/kava/cmd/rocksdb/compact.go

```diff
@@ -0,0 +1,170 @@
+//go:build rocksdb
+// +build rocksdb
+
+package rocksdb
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strconv"
+	"syscall"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/server"
+	"github.com/linxGnu/grocksdb"
+	"github.com/spf13/cobra"
+
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+var CompactRocksDBCmd = &cobra.Command{
+	Use:   "compact <state|blockstore>",
+	Short: "force compacts RocksDB",
+	Long: `This is a utility command that performs a force compaction on the state or
+blockstore. This should only be run once the node has stopped.`,
+	Args: cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+
+		clientCtx := client.GetClientContextFromCmd(cmd)
+		ctx := server.GetServerContextFromCmd(cmd)
+
+		if server.GetAppDBBackend(ctx.Viper) != "rocksdb" {
+			return errors.New("compaction is currently only supported with rocksdb")
+		}
+
+		if args[0] != "state" && args[0] != "blockstore" {
+			return errors.New("invalid db name, must be either 'state' or 'blockstore'")
+		}
+
+		compactRocksDBs(clientCtx.HomeDir, logger, args[0])
+		return nil
+	},
+}
+
+// compactRocksDBs performs a manual compaction on the given db.
+func compactRocksDBs(
+	rootDir string,
+	logger log.Logger,
+	dbName string,
+) {
+	options := grocksdb.NewDefaultOptions()
+
+	dbPath := filepath.Join(rootDir, "data", dbName+".db")
+	logger.Info("opening db", "path", dbPath)
+
+	store, err := grocksdb.OpenDb(options, dbPath)
```

Comment:
We'll have to extend Yevhenii's work at https://github.com/Kava-Labs/kava/blob/master/cmd/kava/opendb/opendb_rocksdb.go#L92 to separate returning of the db instance from the dbm wrapper.

Otherwise, we'll end up opening the db with possibly incompatible options and/or the compacted files will be re-written with different block sizes, etc than app.toml specifies

Reply:
Edit: I see now this is for the state & blockstore -- those use default options from tmdb -- I"m not sure if the default grocksdb options match. The app.toml settings will only apply applicationdb.

On further thought, maybe we should just load the latest options from the DB and ignore defaults or reading app.toml.

Grocksdb can do that like so - https://github.com/Kava-Labs/kava/blob/master/cmd/kava/opendb/opendb_rocksdb.go#L118.  So we always open & compact with the options the db was last opened with

Reply:
oh great, useful function, will update options used